File: Chapter.04/DataParallelism.Part1.CSharp/Mandelbrot.cs
using System;
using System.Threading.Tasks;
using SixLabors.ImageSharp;
using SixLabors.ImageSharp.Advanced;
using SixLabors.ImageSharp.PixelFormats;

namespace DataParallelism.Part1.CSharp
{
    public class Mandelbrot
    {
        private const float Width = 2.5f;
        private const float Height = 2.5f;

        private static readonly Complex Center = new Complex(-0.75f, 0.0f);

        private static readonly Rgba32 black = new Rgba32(0, 0, 0, byte.MaxValue);
        private static readonly Rgba32 white = new Rgba32(byte.MaxValue, byte.MaxValue, byte.MaxValue, byte.MaxValue);


        private static readonly ComplexStruct _centerStruct = new ComplexStruct(-0.75f, 0.0f);

        private static float ComputeRow(int col, int cols)
        {
            return Center.Real - Width / 2.0f + col * Width / cols;
        }

        private static float ComputeColumn(int row, int rows)
        {
            return Center.Imaginary - Height / 2.0f + row * Height / rows;
        }

        public static Image SequentialMandelbrot(int size)
        {
            var rows = size;
            var cols = size;

            var image = new Image<Rgba32>(rows, cols);

            // Listing 4.2 Sequential Mandelbrot
            Func<Complex, int, bool> isMandelbrot = (complex, iterations) => //#A
            {
                var z = new Complex(0.0f, 0.0f);
                var acc = 0;
                while (acc < iterations && z.Magnitude < 2.0)
                {
                    z = z * z + complex;
                    acc += 1;
                }

                return acc == iterations;
            };

            for (var col = 0; col < cols; col++) //#B
            {
                var pixelRowSpan = image.GetPixelRowSpan(col);

                for (var row = 0; row < rows; row++) //#B
                {
                    var x = ComputeRow(row, rows); //#C
                    var y = ComputeColumn(col, cols); //#C
                    var c = new Complex(x, y);
                    var color = isMandelbrot(c, 100) ? black : white; //#D

                    pixelRowSpan[row] = color; //#E
                }
            }

            // you can save the image using the ".Save" API (Ex.  image.Save("./myImage.jpg");
            return image;
        }

        public static Image ParallelMandelbrot(int size)
        {
            var rows = size;
            var cols = size;

            var image = new Image<Rgba32>(rows, cols);

            // Listing 4.3 Parallel Mandelbrot
            Func<Complex, int, bool> isMandelbrot = (complex, iterations) =>
            {
                var z = new Complex(0.0f, 0.0f);
                var acc = 0;
                while (acc < iterations && z.Magnitude < 2.0)
                {
                    z = z * z + complex;
                    acc += 1;
                }

                return acc == iterations;
            };

            Parallel.For(0, cols - 1, col =>
            {
                var pixelRowSpan = image.GetPixelRowSpan(col);
                //#A
                for (var row = 0; row < rows; row++)
                {
                    var x = ComputeRow(row, rows); //#C
                    var y = ComputeColumn(col, cols); //#C
                    var c = new Complex(x, y);
                    var color = isMandelbrot(c, 100) ? black : white; //#D

                    pixelRowSpan[row] = color; //#E
                }
            });

            return image;
        }

        public static Image ParallelMandelbrotOversaturation(int size)
        {
            var rows = size;
            var cols = size;

            var image = new Image<Rgba32>(rows, cols);

            Func<Complex, int, bool> isMandelbrot = (complex, iterations) =>
            {
                var z = new Complex(0.0f, 0.0f);
                var acc = 0;
                while (acc < iterations && z.Magnitude < 2.0)
                {
                    z = z * z + complex;
                    acc += 1;
                }

                return acc == iterations;
            };

            Parallel.For(0, cols - 1, col =>
            {
                var pixelRowSpan = new Memory<Rgba32>(image.GetPixelRowSpan(col).ToArray());

                Parallel.For(0, rows - 1, row =>
                {
                    var span = pixelRowSpan.Span;
                    var x = ComputeRow(row, rows); //#C
                    var y = ComputeColumn(col, cols); //#C
                    var c = new Complex(x, y);
                    var color = isMandelbrot(c, 100) ? black : white; //#D

                    span[row] = color; //#E
                });
            });

            return image;
        }

        private static float ComputeRowStruct(int col, int cols)
        {
            return _centerStruct.Real - Width / 2.0f + col * Width / cols;
        }

        private static float ComputeColumnStruct(int row, int rows)
        {
            return _centerStruct.Imaginary - Height / 2.0f + row * Height / rows;
        }

        public static Image ParallelStructMandelbrot(int size)
        {
            var rows = size;
            var cols = size;

            var image = new Image<Rgba32>(rows, cols);

            Func<ComplexStruct, int, bool> isMandelbrot = (complex, iterations) =>
            {
                var z = new ComplexStruct(0.0f, 0.0f);
                var acc = 0;
                while (acc < iterations && z.Magnitude < 2.0)
                {
                    z = z * z + complex;
                    acc += 1;
                }

                return acc == iterations;
            };

            Parallel.For(0, cols - 1, col =>
            {
                var pixelRowSpan = image.GetPixelRowSpan(col);
                for (var row = 0; row < rows; row++)
                {
                    var x = ComputeRowStruct(row, rows);
                    var y = ComputeColumnStruct(col, cols);
                    var c = new ComplexStruct(x, y);
                    var color = isMandelbrot(c, 100) ? black : white; //#D

                    pixelRowSpan[row] = color; //#E
                }
            });
            return image;
        }
    }
}

File: Chapter.04/DataParallelism.Part1.FSharp/Mandelbrot.fs
namespace DataParallelism.Part1.FSharp

module Mandelbrot =

    open System
    open System.Linq

    open SixLabors.ImageSharp
    open SixLabors.ImageSharp.Advanced
    open SixLabors.ImageSharp.PixelFormats
    
    open System.Collections.Concurrent
    open FSharp.Collections.ParallelSeq

    // Listing 4.1 Complex number object
    type Complex(real : float, imaginary : float) =
        member this.Real = real
        member this.Imaginary = imaginary

        member this.Magnitude =
            sqrt(this.Real * this.Real + this.Imaginary * this.Imaginary)

        static member (+) (c1 : Complex, c2 : Complex) =
            new Complex(c1.Real + c2.Real, c1.Imaginary + c2.Imaginary)
        static member (*) (c1 : Complex, c2 : Complex) =
            new Complex(c1.Real * c2.Real - c1.Imaginary * c2.Imaginary,
                        c1.Real * c2.Imaginary + c1.Imaginary * c2.Real)

    let black = new Rgba32((byte) 0, (byte) 0, (byte) 0, Byte.MaxValue)
    let white = new Rgba32(Byte.MaxValue, Byte.MaxValue, Byte.MaxValue, Byte.MaxValue)

    let sequentialMandelbrot size =
        let rows = size
        let cols = size
        
        let center = Complex(-0.75, 0.0)
        let width, height = 2.5, 2.5

        let colToX col = center.Real - width / 2.0 +
                         float(col) * width / float(cols)
        let rowToY row = center.Imaginary - height / 2.0 +
                         float(row) * height / float(rows)

        let isMandelbrot c iterations =
            let rec isMandelbrot (z:Complex) (rep:int) =
                if rep < iterations && z.Magnitude < 2.0 then
                    isMandelbrot (z * z + c) (rep + 1)
                else rep = iterations
            isMandelbrot c 0

        let image = new Image<Rgba32>(rows, cols)
        

        for col = 0 to cols - 1 do
          let pixelRowSpan = image.GetPixelRowSpan(col)
          for  row = 0 to rows - 1 do
            let x,y = colToX row, rowToY col
            let c = Complex(x, y)
            
            let color =
                if isMandelbrot c 100 then black else white
                
            pixelRowSpan.[row] <- color //#E
           
        image


    let parallelMandelbrot size =
        let rows = size
        let cols = size
        
        let center = Complex(-0.75, 0.0)
        let width, height = 2.5, 2.5

        let colToX col = center.Real - width / 2.0 +
                         float(col) * width / float(cols)
        let rowToY row = center.Imaginary - height / 2.0 +
                         float(row) * height / float(rows)

        let isMandelbrot c iterations =
            let rec isMandelbrot (z:Complex) (rep:int) =
                if rep < iterations && z.Magnitude < 2.0 then
                    isMandelbrot (z * z + c) (rep + 1)
                else rep = iterations
            isMandelbrot c 0

        let image = new Image<Rgba32>(rows, cols)
        
        Partitioner.Create(0, cols - 1).AsParallel()
        |> PSeq.withDegreeOfParallelism(Environment.ProcessorCount)
        |> PSeq.withMergeOptions(ParallelMergeOptions.FullyBuffered)
        |> PSeq.iter(fun (s,e) ->
            for col = s to e - 1 do          
              let pixelRowSpan = image.GetPixelRowSpan(col)          
              for  row = 0 to rows - 1 do
                let x,y = colToX row, rowToY col
                let c = Complex(x, y)
                
                let color =
                    if isMandelbrot c 100 then black else white
                    
                pixelRowSpan.[row] <- color //#E
           )
        image




    [<Struct>]
    type ComplexStruct(real : float, imaginary : float) =
        member this.Real = real
        member this.Imaginary = imaginary

        member this.Magnitude =
            sqrt(this.Real * this.Real + this.Imaginary * this.Imaginary)

        static member (+) (c1 : ComplexStruct, c2 : ComplexStruct) =
            new ComplexStruct(c1.Real + c2.Real, c1.Imaginary + c2.Imaginary)
        static member (*) (c1 : ComplexStruct, c2 : ComplexStruct) =
            new ComplexStruct(c1.Real * c2.Real - c1.Imaginary * c2.Imaginary,
                        c1.Real * c2.Imaginary + c1.Imaginary * c2.Real)

    let parallelMandelbrotStruct size =
        let rows = size
        let cols = size
        
        let center = ComplexStruct(-0.75, 0.0)
        let width, height = 2.5, 2.5

        let colToX col = center.Real - width / 2.0 +
                         float(col) * width / float(cols)
        let rowToY row = center.Imaginary - height / 2.0 +
                         float(row) * height / float(rows)

        let isMandelbrot c iterations =
            let rec isMandelbrot (z:ComplexStruct) (rep:int) =
                if rep < iterations && z.Magnitude < 2.0 then
                    isMandelbrot (z * z + c) (rep + 1)
                else rep = iterations
            isMandelbrot c 0

        let image = new Image<Rgba32>(rows, cols)
        
        Partitioner.Create(0, cols - 1).AsParallel()
        |> PSeq.withDegreeOfParallelism(Environment.ProcessorCount)
        |> PSeq.withMergeOptions(ParallelMergeOptions.FullyBuffered)
        |> PSeq.iter(fun (s,e) ->
            for col = s to e - 1 do
              let pixelRowSpan = image.GetPixelRowSpan(col) 
              for  row = 0 to rows - 1 do
                let x,y = colToX row, rowToY col
                let c = ComplexStruct(x, y)
                
                let color =
                    if isMandelbrot c 100 then black else white
                    
                pixelRowSpan.[row] <- color //#E
           )
        image
